Teach PkgServer to serve multiple registry flavors

staticfloat opened the pull request:
The storage server now exports two different "flavors" of registry
declarations, `eager` and `conservative`:

- `eager` registries are published as soon as possible, and should track
upstream git repositories quite closely, (in general, it should be less
than two minutes behind the state of the upstream git repository).  The
downside is that the storage server may not have processed some of the
resources in the registry, so if you are a user behind a restrictive
firewall and cannot download resources off of arbitrary servers, we
suggest you use the `conservative` registry.

- `conservative` registries are published only once the StorageServer
has at least tried to process all resources (packages, artifacts,
etc...) within a registry.  It does not guarantee that all resources are
downloadable (after all, there are some package versions that contain
links to artifacts that simply do not exist) but it gives a "best
effort" treehash to use.  The downside is that in the event that a large
number of new resources have been added to the registry recently, it may
take a little while for the storage server to process them all, so the
`conservative` registry flavor may lag behind real-time by a few hours.

Pkg clients that ask for `/registries` are, by default, served a
`conservative` registry, unless one of the following two conditions
hold:

- The Pkg client declares itself to be running CI (via one of the CI
  variables transmitted in the `Julia-Ci-Variables` header, see [0] for
  more), in which case we assume you likely want `eager`.

- The Pkg client transmits a `Julia-Registry-Preference` header set to
  one of the valid flavors, which currently includes `eager` and
  `conservative`.

Now that [1] has been merged into `Pkg.jl`, users can easily set their
own preference header by simply defining the environment mapping
`JULIA_PKG_SERVER_REGISTRY_PREFERENCE=eager`.

[0] https://github.com/JuliaLang/Pkg.jl/blob/bb180de2802a8320241e10721c1ac4eefd3eb27d/src/PlatformEngines.jl#L200-L213

[1] JuliaLang/Pkg.jl#2770

staticfloat committed:
Teach PkgServer to serve multiple registry flavors

The storage server now exports two different "flavors" of registry
declarations, `eager` and `conservative`:

- `eager` registries are published as soon as possible, and should track
upstream git repositories quite closely, (in general, it should be less
than two minutes behind the state of the upstream git repository).  The
downside is that the storage server may not have processed some of the
resources in the registry, so if you are a user behind a restrictive
firewall and cannot download resources off of arbitrary servers, we
suggest you use the `conservative` registry.

- `conservative` registries are published only once the StorageServer
has at least tried to process all resources (packages, artifacts,
etc...) within a registry.  It does not guarantee that all resources are
downloadable (after all, there are some package versions that contain
links to artifacts that simply do not exist) but it gives a "best
effort" treehash to use.  The downside is that in the event that a large
number of new resources have been added to the registry recently, it may
take a little while for the storage server to process them all, so the
`conservative` registry flavor may lag behind real-time by a few hours.

Pkg clients that ask for `/registries` are, by default, served a
`conservative` registry, unless one of the following two conditions
hold:

- The Pkg client declares itself to be running CI (via one of the CI
  variables transmitted in the `Julia-Ci-Variables` header, see [0] for
  more), in which case we assume you likely want `eager`.

- The Pkg client transmits a `Julia-Registry-Preference` header set to
  one of the valid flavors, which currently includes `eager` and
  `conservative`.

Now that [1] has been merged into `Pkg.jl`, users can easily set their
own preference header by simply defining the environment mapping
`JULIA_PKG_SERVER_REGISTRY_PREFERENCE=eager`.

[0] https://github.com/JuliaLang/Pkg.jl/blob/bb180de2802a8320241e10721c1ac4eefd3eb27d/src/PlatformEngines.jl#L200-L213

[1] https://github.com/JuliaLang/Pkg.jl/pull/2770

staticfloat committed:
Fix comparison

DilumAluthge commented:
Does condition 2 override condition 1? E.g. if the client declares that it is running on CI, but also provides the `JULIA_PKG_SERVER_REGISTRY_PREFERENCE=conservative` header, which registry will it receive?

staticfloat committed:
Update `registry_lag.jl` to show both flavors

This way we can easily check whether the `conservative` flavor is
significantly behind the `eager` flavor.

staticfloat commented:
Yes, `JULIA_PKG_SERVER_REGISTRY_PREFERENCE`/`Julia-Registry-Preference` takes precedence over all other considerations.

johnnychen94 commented:
Just a quick question, does this solve the pkg server latency issue: https://github.com/JuliaRegistries/General/issues/16777?

staticfloat commented:
Yes, this is one part of it

StefanKarpinski commented:
My gut says we should do this with a redirect based on the header instead of serving different content to different people for `/registries`. That is, when someone asks for `/registries` it’s a redirect to one of `/registries.eager` or `/registries.conservative` which is then served. I know those aren’t content addressed, but that way they could each be cached for some amount of time and have the correct behavior for clients with different headers. Maybe it doesn’t matter though.

staticfloat committed:
Switch `/registries` to serving a 302

fredrikekre commented:
Why would you want new releases faster on CI than to your own computer?

StefanKarpinski commented:
CI is often run very shortly after a release is made, which breaks if the new version isn't available.

fredrikekre commented:
Most of the complaints I have seen about registry delays have not been for CI though. But why is CI run shortly after a release usually? Are people monitoring the registry repo and then trigger CI for their other package that suddenly now requires the absolutely latest release? Wouldn't you first want to develop against the new release, run tests locally and then push? That would suggest you want the latest release locally first though. Maybe just a workflow I don't understand.

giordano commented:
> Most of the complaints I have seen about registry delays have not been for CI though.

Probably because most CI jobs [are not using PkgServer at all](https://github.com/julia-actions/julia-runtest/blob/a5f2948fcb2fe71a428a722c286a48efd45bc610/action.yml#L33-L48) for this precise reason.

fredrikekre commented:
Right, but still, there are complaints about it for local use, so I would prefer that eager was default. For everyone that is not behind a firewall that is good. And if you are behind a firewall and need to update settings there you might as well also set this flag.

fredrikekre commented:
Or perhaps put it like this; I see more complaints about the registry not being eager than complaints about it not being conservative.

StefanKarpinski commented:
We've had this exact discussion like a dozen times:

* Package developers want an eager registry. These are engaged, vocal users who we are very likely to hear from when they encounter issues. When they do complain about not getting an eager registry, we can tell them to do `export JULIA_PKG_SERVER_REGISTRY_PREFERENCE=eager` and they'll get an eager registry. Problem solved forever. This population is at most 0.1% of Julia users. 

* People behind a strict firewall want a complete registry. This is a quiet, not very vocal group of users. They mostly won't complain (or even know to) and will just silently conclude that Julia's package manager is broken and shitty. This population is probably about 5% of Julia users.

* The other 94.9% of Julia users don't care if they get an eager or a complete registry as long as they get new package versions in a reasonably timely fashion.

It might be fine to default to eager if there was some clever way to figure out in the Pkg client that a strict firewall is causing problems and either change the default when that happens or print a helpful warning, but that's not simple to do and even if we do something like that, it won't help users of existing Julia versions in the wild, whereas this will.

StefanKarpinski commented:
@staticfloat, a reason to do the redirect approach is that it makes it easy to tell which you're getting: do `curl -iL https://pkg.julialang.org` and see what end-point you get redirected to.

fredrikekre commented:
My point is that with this default everyone in the 5% people that care now have to configure something (firewall or `JULIA_PKG_SERVER_REGISTRY_PREFERENCE`), but with eager as the default only the people that have to configure things anyway will have to make another setting.

For the 95% of other users, I am sure they would be happy to have access to new releases faster. In addition, with conservative as default they will now get slower updates as compared to before, right?

StefanKarpinski commented:
> My point is that with this default everyone in the 5% people that care now have to configure something 

No, only the 0.1% of users who need an eager registry have to do anything. And those are precisely the users who are best positioned to learn that they need to do this.

> For the 95% of other users, I am sure they would be happy to have access to new releases faster.

Based on what? Why would they care? I'm in this group and I don't care at all about waiting a couple of hours every now and then to get some package update. I update my packages very irregularly anyway.

You're not proposing any solutions for the 5% of people behind firewalls for whom an eager default is broken. How do we fix that for them or communicate to them what they need to do to fix the problem?

johnnychen94 commented:
My own experience in watching discussions in the JuliaCN community is that people would throw something like https://github.com/tuna/issues/issues/866 to everywhere except to the Pkg.jl. Making eager the default would trigger a lot of such issues again.

fredrikekre commented:
I personally don't care and have never really been bothered by this, I am just saying that it seems like people care since there is even a [pinned discourse post](https://discourse.julialang.org/t/general-registry-delays-and-a-workaround/67537) describing how to get the eager behavior, and quoting from that post:

> It is a known issue, and we are actively working on it.

And now the solution to that issue is to make it worse (i.e. slower)?

> You're not proposing any solutions for the 5% of people behind firewalls for whom an eager default is broken. How do we fix that for them or communicate to them what they need to do to fix the problem?

My solution was to answer the question "How do I use Pkg behind a firewall?" with "You allow pkg.julialang.org and also set this env var".

giordano commented:
> And now the solution to that issue is to make it worse (i.e. slower)?

Maybe I'm missing something but I thought that the `conservative` mode was equivalent to the current behaviour of PkgServer?

johnnychen94 commented:
> You allow pkg.julialang.org and also set this env var

It isn't that simple I think. The problem with the eager mode is that when you try to request resources that the upstream storage server doesn't yet have, the Pkg server would return a (maybe) 404 error and then trigger the fallback solution of Pkg, which is, downloading from, e.g., GitHub and AWS s3. Allowing "github.com" is sometimes impractical for various reasons. For example, for people in China's mainland, the [Great Firewall](https://en.wikipedia.org/wiki/Great_Firewall) regularly drops the connection to GitHub/S3.

fredrikekre commented on src/resource.jl:
```diff
@@ -10,17 +10,17 @@ const resource_re = Regex("""
+function get_registries(server::AbstractString, flavor::AbstractString)
     regs = Dict{String,String}()
-    response = HTTP.get("$server/registries", status_exception = false)
+    response = HTTP.get("$server/registries.$(flavor)", status_exception = false)
```
How does PkgServer know if the storage server have flavors? It would be great to still support unflavored storage servers at least.

DilumAluthge replied:
Maybe the storage server can expose an endpoint that the Pkg server can query to get the list of supported flavors?

staticfloat replied:
Okay, I have added a "flavorless" mode, which is autodetected if all of its upstream storage servers don't respond to these resources.

fredrikekre commented:
> Maybe I'm missing something but I thought that the conservative mode was equivalent to the current behaviour of PkgServer?

No, IIUC, `conservative` will wait until everything is stored. The current state is somewhere in between eager and conservative.

StefanKarpinski commented:
That's a really excellent point, @johnnychen94—some entire *countries* are firewalled from GitHub, GitLab, etc. So increase that estimate from 5% of users behind corporate firewalls to some 15% and add a language barrier making it all the more important that we default to something that just works. There's also the analysis of the severity of the failure modes of these choices:

* Person who prefers eager: they can install packages but might have to wait a few hours for one that's as current as they would like. Slightly annoying, not a big deal.

* Person who gets conservative but needs eager: they cannot install packages at all.

johnnychen94 commented:
> The current state is somewhere in between eager and conservative.

Do you mean the implementation of juliacomputing's storage server broke the protocol described in https://github.com/JuliaLang/Pkg.jl/issues/1377 at some point? Because the [public legacy `gen_static.jl`](https://github.com/JuliaPackaging/PkgServer.jl/blob/3d2b47cefefb33f96a48509853e43aa4b5785c1b/bin/gen_static.jl#L376-L381) is designed to be conservative.

> From https://github.com/JuliaLang/Pkg.jl/issues/1377:
"The primary guarantee that a Storage Server makes is that if it has ever successfully served a resource—registry tree, package source tree, artifact tree—it must be able to serve that same resource version forever."

staticfloat commented:
> "The primary guarantee that a Storage Server makes is that if it has ever successfully served a resource—registry tree, package source tree, artifact tree—it must be able to serve that same resource version forever."

This is still true; the issue is whether you publish a registry tree that references resources that are not stored yet (or ever).

StefanKarpinski commented:
> You allow pkg.julialang.org and also set this env var

Most firewalls already allow pkg.julialang.org by default—GitHub and so on are explicitly disallowed. So if the default registry is conservative then everything just works and nothing at all needs to be done.

staticfloat commented:
> Maybe I'm missing something but I thought that the conservative mode was equivalent to the current behaviour of PkgServer?

In the past, we did something very similar to `conservative`, but then we got complaints (mostly around CI issues) when new LLVM versions came out.  Then we switched to something closer to `eager` (but it would still get hung up every now and then), so now we have a true `eager`, as well as `conservative`.

In my opinion, the philosophy of the PkgServer lends itself to defaulting everyone to conservative; the whole reason we built this thing was to make it easier to get resources for people who cannot necessarily hit the whole internet; the performance improvements are just really large icing on the cake.

I do agree that since CI jobs tend to want the absolute latest version, it makes sense to do something special there.  I'm not sure if the answer is to auto-detect on the server side, or to have our CI harnesses set `Julia-Registry-Preference`, but _something_ should be done.

staticfloat committed:
Support flavorless operation

If your only upstream storage servers are flavorless, switch back to
flavorless mode as well.

staticfloat committed:
Update dependencies and codebase to assume Julia v1.6+

staticfloat committed:
Add tests for flavorless mode, enable configuration of flavors

Note that this does not use the "default" server, so it's never tested
in Docker, but that's fine, as we don't expect this feature to be used
much.

staticfloat committed:
Disable automatic precompilation to speed up tests

staticfloat committed:
Fix `code_dir` definition

StefanKarpinski reviewed (approved):
Seems good to me. I say we give this a shot.

staticfloat commented:
I also think something that will help us all a bit is to get hard data on the time lag between `eager` and `conservative`.  The `test/debugging/registry_lag.jl` script in this repository will now measure that, and we may find that the two are (99% of the time, at least) exactly identical.  I'd love it if someone can collect this data over time and then plot it for us (over, say, a month) just so we can get an idea of how important the distinction between the two kinds are, since right now we have an intuition that sometimes it's bad, but perhaps with all the optimization of the storage server I've done over the past week, we don't need to worry about it anymore.

StefanKarpinski commented:
There's also the future possibility if there are occasional bursts of processing work that need to be done for expensive registry updates that we could spin up additional machine to help with the workload. Spinning up machines takes some time, but not more than a couple of minutes, so basically any size of registry update could potentially be processed in no more than ten minutes by distributing the work.
